File: app/src/main/java/com/fastaccess/ui/adapter/AssigneesAdapter.java
package com.fastaccess.ui.adapter;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.ViewGroup;

import com.fastaccess.data.dao.model.User;
import com.fastaccess.ui.adapter.viewholder.AssigneesViewHolder;
import com.fastaccess.ui.widgets.recyclerview.BaseRecyclerAdapter;
import com.fastaccess.ui.widgets.recyclerview.BaseViewHolder;

import java.util.List;

/**
 * Created by Kosh on 11 Nov 2016, 2:07 PM
 */

public class AssigneesAdapter extends BaseRecyclerAdapter<User, AssigneesViewHolder, BaseViewHolder.OnItemClickListener<User>> {

    public interface OnSelectAssignee {
        boolean isAssigneeSelected(int position);

        void onToggleSelection(int position, boolean select);
    }

    private final OnSelectAssignee onSelectAssignee;

    public AssigneesAdapter(@NonNull List<User> data, @Nullable OnSelectAssignee onSelectAssignee) {
        super(data);
        this.onSelectAssignee = onSelectAssignee;
    }

    @Override protected AssigneesViewHolder viewHolder(ViewGroup parent, int viewType) {
        return AssigneesViewHolder.newInstance(parent, onSelectAssignee, this);
    }

    @Override protected void onBindView(AssigneesViewHolder holder, int position) {
        holder.bind(getItem(position));
    }
}

File: app/src/main/java/com/fastaccess/github/ui/adapter/AssigneesAdapter.kt
package com.fastaccess.github.ui.adapter

import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import com.fastaccess.data.model.ShortUserModel
import com.fastaccess.github.ui.adapter.viewholder.AssigneeViewHolder

/**
 * Created by Kosh on 20.01.19.
 */
class AssigneesAdapter(
    val selection: HashSet<ShortUserModel>,
    val deselection: HashSet<ShortUserModel>
) : ListAdapter<ShortUserModel, AssigneeViewHolder>(DIFF_CALLBACK) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AssigneeViewHolder = AssigneeViewHolder(parent).apply {
        itemView.let { view ->
            view.setOnClickListener {
                val position = adapterPosition
                val item = getItem(position)
                if (!selection.contains(item)) {
                    selection.add(item)
                    deselection.remove(item)
                } else {
                    deselection.add(item)
                    selection.remove(item)
                }
                notifyItemChanged(position)
            }
        }
    }

    override fun onBindViewHolder(holder: AssigneeViewHolder, position: Int) {
        val item = getItem(position)
        holder.bind(item, selection.contains(item))
    }

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<ShortUserModel?>() {
            override fun areItemsTheSame(oldItem: ShortUserModel, newItem: ShortUserModel): Boolean = oldItem.name == newItem.name
            override fun areContentsTheSame(oldItem: ShortUserModel, newItem: ShortUserModel): Boolean = oldItem == newItem
        }
    }
}
